File: supa_rest_api.py
import os
import requests as r
import json


PROJECT_URL  = os.environ.get('SUPABASE_URL',None)
SUPABASE_KEY = os.environ.get('SUPABASE_KEY', None)


class Client:

    def __init__(self, project_url, supabase_key):

        if not SUPABASE_KEY:
            print("Please set up your SUPABASE_KEY environment variable")
            return

        self.project_url  = project_url
        self.supabase_key = supabase_key
        self.base_path    = self.project_url + "/rest/v1"
        self.pg_range     = None
        self.headers      = {'apiKey': self.supabase_key}


    def connect(self, path):
        res = r.get(path, headers = self.headers)
        return (res.status_code, json.dumps(res.json(), indent = 4))


    def pagination(self, start, end):
        self.pg_range = range(start, end)


    def paginationRequested(self):
        self.headers["Range"] = f'{self.pg_range.start}-{self.pg_range.stop}'


    def resultSet(self, path, pagination=False):
        if pagination:
            self.paginationRequested()

        response_code, response_data = self.connect(path)

        #if response_code == 200:
        print(response_data)


    def selectAll(self, table_name, pagination=False):
        path = f'{self.base_path}/{table_name}?select=*'
        self.headers["Authorization"] = f'Bearer {self.supabase_key}'
        self.resultSet(path, pagination)


    def selectColumns(self, table_name, pagination=False, cols=[]):
        csc  = ",".join(cols)
        path = f'{self.base_path}/{table_name}?select={csc}'
        self.headers["Authorization"] = f'Bearer {self.supabase_key}'
        self.resultSet(path, pagination)


def main():
    client = Client(PROJECT_URL, SUPABASE_KEY)
    #client.pagination(0,0)
    client.selectAll("products")
    #client.selectColumns("products", ["id"])


if __name__ == '__main__':
    main()

File: supa_rest_api.js
const https = require('https');

const PROJECT_URL = 'https://utawmowgglsndawzxebx.supabase.co';
const SUPABASE_KEY = process.env.SUPABASE_KEY;


class Client {

  constructor(projectUrl, supabaseKey) {
    if (!supabaseKey) {
      console.error("Please set up your SUPABASE_KEY environment variable");
      return;
    }
    this.projectUrl = projectUrl;
    this.supabaseKey = supabaseKey;
    this.basePath = this.projectUrl + "/rest/v1";
    this.pgRange = null;
    this.headers = {
      'apiKey': this.supabaseKey
    };
  }

  connect(path) {
    return new Promise((resolve, reject) => {
      const options = {
        hostname: this.projectUrl.split('//')[1], // Extract hostname from URL
        path: path,
        method: 'GET',
        headers: this.headers
      };

      const req = https.request(options, (res) => {
        let data = '';
        res.on('data', (chunk) => {
          data += chunk;
        });

        res.on('end', () => {
          try {
            const response = JSON.parse(data);
            resolve({ statusCode: res.statusCode, data: response });
          } catch (error) {
            reject(error);
          }
        });
      });

      req.on('error', reject);
      req.end();
    });
  }

  pagination(start, end) {
    this.pgRange = Array.from({ length: end - start + 1 }, (_, i) => start + i); // Generate inclusive range
  }

  paginationRequested() {
    if (this.pgRange) {
      this.headers['Range'] = `${this.pgRange[0]}-${this.pgRange[this.pgRange.length - 1]}`; // Use last element for end
    }
  }

  async resultSet(path, pagination = false) {
    if (pagination) {
      this.paginationRequested();
    }

    try {
      const response = await this.connect(path);
      console.log(response.data);
    } catch (error) {
      console.error(error);
    }
  }

  selectAll(table_name, pagination = false) {
    const path = `${this.basePath}/${table_name}?select=*`;
    this.headers['Authorization'] = `Bearer ${this.supabaseKey}`;
    this.resultSet(path, pagination);
  }

  selectColumns(table_name, pagination = false, cols = []) {
    const csc = cols.join(',');
    const path = `${this.basePath}/${table_name}?select=${csc}`;
    this.headers['Authorization'] = `Bearer ${this.supabaseKey}`;
    this.resultSet(path, pagination);
  }
}


function main() {
  const client = new Client(PROJECT_URL, SUPABASE_KEY);
  client.selectAll("products");
}


main()
